### ISSUE TITLE
A git push operation is only executed if a target changed in the current pipeline

### ISSUE THREAD

**Issue by olblak** [updatecli/updatecli#869]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

A git push operation is only executed if a target changed in the current pipeline.
This means that if something goes wrong and we have to re-execute Updatecli.

Then it won't push the changes on the remote repository 

### Expected Behavior

Updatecli must idempotent and should compare the upstream branch and the local one to know if it should execute a push operation

### Environment

```markdown
- OS:Linux
- updatecli:0.33.0
```

**Comment by dduportal** [comment]
> 

🤔 Not sure to see why is it a bug? 

If "something goes wrong" then it means it's a bug, could you clarify?

**Comment by lemeurherve** [comment]
Not sure to understand either, if something goes wrong and the push isn't
done, nothing has changed in the repo yhus updatecli should find an update
to perform when reexecuted.

Le ven. 16 sept. 2022, 08:52, Damien Duportal ***@***.***> a
écrit :

**Comment by olblak** [comment]
In my opinion, It a bug that can be easily worked around by deleting the temporary directory before running `updatecli apply` 
Let me provide an example, in case it's easier to understand: 

First pipeline run:

* `updateci apply --config updatecli.yaml`
  * Detect a change
  * Commit the change 
  * Fail to push the change due to network error

Updatecli user notices that something went wrong so he tries to re-execute Updatecli:
* Reuse previous temporary directory
* No change detected as the local directory still contains the commit from the previous run
* Updatecli skip the git push operation

The way I discovered this behavior was while trying to understand why this repository wasn't updated 
-> https://github.com/epinio/docs/pull/171/files
Basically the script called from Updatecli was also executed in DRY_MODE.
So `updatecli diff` would apply the change without running git push operation because it's in dry_mode then `updatecli apply` would never push the changes because the working directory had already been changed in the `updatecli diff` command

The fix that I want to apply is:
Instead of running git push operation if something changed during the current Updatecli run
I want to diff the local branch and remote branch and push if it's different

**Comment by dduportal** [comment]
🤔 that feels really weird to me to reuse the same temp. directory for the "target" part. Not sure why?

I understand that caching the git repository to ensure you do not clone it on each run make sense.
But your idempotency will never be achievable if you cache the "target" part.

WDYT about the following?

- Before the target, force create the local branch from either:
  - The remote branch (with same name) if exists, to retrieve remote changes
  - Or the "principal" branch (or PR specied destination branch) if no remote branche exist.

=> that would solve your issue

**Comment by dduportal** [comment]
Additionnaly, we might be interested into looking into the behavior of https://github.com/peter-evans/create-pull-request.

For instance (from https://github.com/jenkins-infra/docker-jenkins-lts/actions/runs/3066423725/jobs/4951579291):

```console
2022-09-16T08:23:04.2326842Z ##[group]Run peter-evans/create-pull-request@v4
2022-09-16T08:23:04.2327251Z with:
2022-09-16T08:23:04.2327641Z   commit-message: feat(deps): bump jenkins lts to 2.361.1
2022-09-16T08:23:04.2328059Z   signoff: false
2022-09-16T08:23:04.2328426Z   title: feat(deps): bump jenkins lts to 2.361.1
2022-09-16T08:23:04.2328817Z   branch: latest-lts
2022-09-16T08:23:04.2329463Z   token: ***
2022-09-16T08:23:04.2329838Z   committer: GitHub <noreply@github.com>
2022-09-16T08:23:04.2330310Z   author: garethjevans <garethjevans@users.noreply.github.com>
2022-09-16T08:23:04.2330770Z   delete-branch: false
2022-09-16T08:23:04.2331355Z   body: Automated changes by [create-pull-request](https://github.com/peter-evans/create-pull-request) GitHub action
2022-09-16T08:23:04.2332000Z   draft: false
2022-09-16T08:23:04.2332327Z ##[endgroup]
2022-09-16T08:23:04.3267182Z ##[group]Save persisted git credentials
2022-09-16T08:23:04.3293331Z [command]/usr/bin/git config --local --name-only --get-regexp http.https://github.com/.extraheader ^AUTHORIZATION:
2022-09-16T08:23:04.3406689Z http.https://github.com/.extraheader
2022-09-16T08:23:04.3438351Z [command]/usr/bin/git config --local --get-regexp http.https://github.com/.extraheader ^AUTHORIZATION:
2022-09-16T08:23:04.3474256Z http.https://github.com/.extraheader AUTHORIZATION: basic ***
2022-09-16T08:23:04.3494936Z [command]/usr/bin/git config --local --unset http.https://github.com/.extraheader ^AUTHORIZATION:
2022-09-16T08:23:04.3591313Z Unset config key 'http.https://github.com/.extraheader'
2022-09-16T08:23:04.3592385Z ##[endgroup]
2022-09-16T08:23:04.3593282Z ##[group]Determining the base and head repositories
2022-09-16T08:23:04.3600313Z [command]/usr/bin/git config --local --get remote.origin.url
2022-09-16T08:23:04.3631969Z https://github.com/jenkins-infra/docker-jenkins-lts
2022-09-16T08:23:04.3654815Z ##[endgroup]
2022-09-16T08:23:04.3655451Z Pull request branch target repository set to jenkins-infra/docker-jenkins-lts
2022-09-16T08:23:04.3656265Z ##[group]Configuring credential for HTTPS authentication
2022-09-16T08:23:04.3658033Z [command]/usr/bin/git config --local http.https://github.com/.extraheader AUTHORIZATION: basic ***
2022-09-16T08:23:04.3707819Z ##[endgroup]
2022-09-16T08:23:04.3708468Z ##[group]Checking the base repository state
2022-09-16T08:23:04.3717793Z [command]/usr/bin/git symbolic-ref HEAD --short
2022-09-16T08:23:04.3749694Z main
2022-09-16T08:23:04.3755071Z Working base is branch 'main'
2022-09-16T08:23:04.3759650Z [command]/usr/bin/git remote prune origin
2022-09-16T08:23:04.5704562Z ##[endgroup]
2022-09-16T08:23:04.5710125Z Pull request branch to create or update set to 'latest-lts'
2022-09-16T08:23:04.5711397Z ##[group]Configuring the committer and author
2022-09-16T08:23:04.5712136Z Configured git committer as 'GitHub <noreply@github.com>'
2022-09-16T08:23:04.5713165Z Configured git author as 'garethjevans <garethjevans@users.noreply.github.com>'
2022-09-16T08:23:04.5714063Z ##[endgroup]
2022-09-16T08:23:04.5715051Z ##[group]Create or update the pull request branch
2022-09-16T08:23:04.5715966Z [command]/usr/bin/git symbolic-ref HEAD --short
2022-09-16T08:23:04.5739187Z main
2022-09-16T08:23:04.5742893Z Working base is branch 'main'
2022-09-16T08:23:04.5754083Z [command]/usr/bin/git checkout --progress -B e0d82f06-f0a2-4e28-9109-c86c210e095b HEAD --
2022-09-16T08:23:04.5793175Z Switched to a new branch 'e0d82f06-f0a2-4e28-9109-c86c210e095b'
2022-09-16T08:23:04.5804533Z [command]/usr/bin/git status --porcelain -unormal --
2022-09-16T08:23:04.5843828Z [command]/usr/bin/git diff --quiet --
2022-09-16T08:23:04.5885545Z [command]/usr/bin/git diff --quiet --staged --
2022-09-16T08:23:04.5912512Z [command]/usr/bin/git reset --hard
2022-09-16T08:23:04.5947527Z HEAD is now at b5eb41f chore(deps): update plugins (#574)
2022-09-16T08:23:04.5956708Z [command]/usr/bin/git clean -f -d
2022-09-16T08:23:04.6010049Z Resetting working base branch 'main'
2022-09-16T08:23:04.6014973Z [command]/usr/bin/git checkout --progress main --
2022-09-16T08:23:04.6047660Z Switched to branch 'main'
2022-09-16T08:23:04.6049393Z Your branch is up to date with 'origin/main'.
2022-09-16T08:23:04.6061295Z [command]/usr/bin/git reset --hard origin/main
2022-09-16T08:23:04.6097157Z HEAD is now at b5eb41f chore(deps): update plugins (#574)
2022-09-16T08:23:04.6110553Z [command]/usr/bin/git -c protocol.version=2 fetch --no-tags --progress --no-recurse-submodules --unshallow --force origin latest-lts:refs/remotes/origin/latest-lts
2022-09-16T08:23:04.7875488Z fatal: couldn't find remote ref latest-lts
2022-09-16T08:23:04.7887717Z Pull request branch 'latest-lts' does not exist yet.
2022-09-16T08:23:04.7891734Z [command]/usr/bin/git checkout --progress -B latest-lts e0d82f06-f0a2-4e28-9109-c86c210e095b --
2022-09-16T08:23:04.7954869Z Switched to a new branch 'latest-lts'
2022-09-16T08:23:04.7955754Z [command]/usr/bin/git rev-list --right-only --count main...latest-lts
2022-09-16T08:23:04.7981523Z 0
2022-09-16T08:23:04.7991652Z Branch 'latest-lts' is not ahead of base 'main' and will not be created
2022-09-16T08:23:04.7992476Z [command]/usr/bin/git rev-parse HEAD
2022-09-16T08:23:04.8015923Z b5eb41f14000f50ed717c2286367007b75c830a9
2022-09-16T08:23:04.8024572Z [command]/usr/bin/git branch --delete --force e0d82f06-f0a2-4e28-9109-c86c210e095b
2022-09-16T08:23:04.8054124Z Deleted branch e0d82f06-f0a2-4e28-9109-c86c210e095b (was b5eb41f).
2022-09-16T08:23:04.8058972Z ##[endgroup]
2022-09-16T08:23:04.8059984Z ##[group]Restore persisted git credentials
2022-09-16T08:23:04.8066348Z [command]/usr/bin/git config --local --name-only --get-regexp http.https://github.com/.extraheader ^AUTHORIZATION:
2022-09-16T08:23:04.8100141Z http.https://github.com/.extraheader
2022-09-16T08:23:04.8110275Z [command]/usr/bin/git config --local --get-regexp http.https://github.com/.extraheader ^AUTHORIZATION:
2022-09-16T08:23:04.8136353Z http.https://github.com/.extraheader AUTHORIZATION: basic ***
2022-09-16T08:23:04.8144981Z [command]/usr/bin/git config --local --unset http.https://github.com/.extraheader ^AUTHORIZATION:
2022-09-16T08:23:04.8172876Z Unset config key 'http.https://github.com/.extraheader'
2022-09-16T08:23:04.8180487Z [command]/usr/bin/git config --local http.https://github.com/.extraheader AUTHORIZATION: basic ***
2022-09-16T08:23:04.8213083Z Persisted git credentials restored
2022-09-16T08:23:04.8218349Z ##[endgroup]
```

**Comment by olblak** [comment]
> Before the target, force create the local branch from either:
> The remote branch (with same name) if exists, to retrieve remote changes
> Or the "principal" branch (or PR specied destination branch) if no remote branche exist.

It's already what we do
* before target, generate a uniq branch name which must be the same across different  pipeline run. For that we generate a hash of the target spec. 
* If the newly generate branch is already on the remote location, then we fetch and merge.

what I would like to implement is something mentioned in the logs you submitted

`2022-09-16T08:23:04.7991652Z Branch 'latest-lts' is not ahead of base 'main' and will not be created`

because right now, all we check is if a target changed something or not. Instead we should check if the local "branch" is ahead of the remote or not.

Now to the question, should we use a "working branch" such as "updatecli_xxxx"?
The git scm doesn't, while github and gitea only support temporary working branch.

The idea behind this PR https://github.com/updatecli/updatecli/pull/745 was to allow to configure either one or the other behavior.
